projects/ADuCM360_demo_cn0411/src/CN0411.c

```diff
@@ -0,0 +1,1876 @@
+		break;
+	case ADC_SINGLE_CONV:
+		ret = CN0411_ADC_ch5 (cn0411_dev, ADC_CH_ENABLE);
+		ret |= CN0411_ADC_operation_mode(cn0411_dev, SINGLE_CONV);
+		if (AD7124_WaitForConvReady(&cn0411_dev->ad7124_dev, ADC_TIMEOUT)
+		    == -3) {
+			printf("TIMEOUT\n");
+			return -1;
+		}
+		ret |= AD7124_ReadRegister(&cn0411_dev->ad7124_dev,
+					   &cn0411_dev->ad7124_dev.regs[AD7124_Data]);
+		ch5_data = cn0411_dev->ad7124_dev.regs[AD7124_Data].value;
+		ret |= CN0411_ADC_ch5(cn0411_dev, ADC_CH_DISABLE);
+		break;
+	default:
+		ret = CN0411_FAILURE;
+		break;
+	}
+	cn0411_dev->read_v_r200s = ch5_data * VREF / 0xFFFFFF;
+
+	return ret;
+}
+
+/**
+ * CN0411 compute input resistance
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_compute_rdres(cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+	float ipp;
+
+	ret = CN0411_read_vpp(cn0411_dev);
+	ipp = (2*cn0411_dev->v_dac - cn0411_dev->vpp)
+	      / cn0411_dev->r_gain[cn0411_dev->ch_gain];
+	cn0411_dev->rdres = cn0411_dev->vpp/ipp;
+	return ret;
+}
+
+/**
+ * CN0411 compute electric conductivity
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_compute_cond(cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+	float ipp, g;
+
+	ret = CN0411_premeasurement(cn0411_dev);
+	ret |= CN0411_read_vpp(cn0411_dev);
+	ipp = (2*cn0411_dev->v_exc - cn0411_dev->vpp)
+	      / cn0411_dev->r_gain[cn0411_dev->ch_gain];
+	cn0411_dev->rdres = cn0411_dev->vpp/ipp - (cn0411_dev->offset_res);
+	g = 1/cn0411_dev->rdres;
+	cn0411_dev->cond = cn0411_dev->cell_const * g;
+
+	return ret;
+}
+
+/**
+ * CN0411 compute offset resistance
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_compute_off_res(cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+	float ipp;
+
+	ret = CN0411_premeasurement(cn0411_dev);
+	ret |= CN0411_read_vpp(cn0411_dev);
+	ipp = (2*cn0411_dev->v_exc - cn0411_dev->vpp)
+	      / cn0411_dev->r_gain[cn0411_dev->ch_gain];
+	cn0411_dev->offset_res = cn0411_dev->vpp/ipp - PREC_REF_RES;
+
+	return ret;
+}
+
+/**
+ * CN0411 temperature compensation of conductivity
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_compensate_cond(cn0411_device *cn0411_dev)
+{
+	int32_t ret;
+
+	ret = CN0411_read_temp(cn0411_dev);
+	ret |= CN0411_compute_cond(cn0411_dev);
+	cn0411_dev->comp_cond = cn0411_dev->cond
+				/ (1 + cn0411_dev->solution.temp_coeff
+				   * (cn0411_dev->temp - TCAL));
+
+	return ret;
+}
+
+/**
+ * CN0411 compute TDS
+ *
+ * @param cn0411_dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+*/
+int32_t CN0411_compute_tds(cn0411_device *cn0411_dev)
+{
+
+	int32_t ret;
+
+	ret = CN0411_compensate_cond(cn0411_dev);
```

Comment:
```
if (ret < 0)
        return ret;
```
In this cases, you do not want to continue, because the variable that is set bellow, will contain an invalid value.